packaging/macos/utility.m

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+
+#import <Cocoa/Cocoa.h>
+#include <dlfcn.h>
+#include <libgen.h>
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/sysctl.h>
+#include <sys/resource.h>
+#include <mach/machine.h>
+
+#define SYSTEM_MONO_PATH @"/Library/Frameworks/Mono.framework/Versions/Current/"
+#define SYSTEM_MONO_MIN_VERSION @"6.4"
+#define DOTNET_MIN_MACOS_VERSION 10.15
+
+typedef void* hostfxr_handle;
+struct hostfxr_initialize_parameters
+{
+	size_t size;
+	char *host_path;
+	char *dotnet_root;
+};
+
+typedef int32_t(*hostfxr_initialize_for_dotnet_command_line_fn)(
+	int argc,
+	char **argv,
+	struct hostfxr_initialize_parameters *parameters,
+	hostfxr_handle *host_context_handle);
+
+typedef int32_t(*hostfxr_run_app_fn)(const hostfxr_handle host_context_handle);
+typedef int32_t(*hostfxr_close_fn)(const hostfxr_handle host_context_handle);
+typedef int (* mono_main)(int argc, char **argv);
+
+int launch_mono(int argc, char **argv, char *modId)
+{
+	NSString *exePath = [[[NSBundle mainBundle] bundlePath] stringByAppendingPathComponent: @"Contents/MacOS/"];
+	NSString *hostPath = [SYSTEM_MONO_PATH stringByAppendingPathComponent: @"lib/libmonosgen-2.0.dylib"];
+	NSString *dllPath = [exePath stringByAppendingPathComponent: @"mono/OpenRA.Utility.dll"];
+
+	// TODO: This snippet increasing the open file limit was copied from
+	// the monodevelop launcher stub. It may not be needed for OpenRA.
+	struct rlimit limit;
+	if (getrlimit(RLIMIT_NOFILE, &limit) == 0 && limit.rlim_cur < 1024)
+	{
+		limit.rlim_cur = limit.rlim_max < 1024 ? limit.rlim_max : 1024;
+		setrlimit(RLIMIT_NOFILE, &limit);
+	}
+
+	void *libmono = dlopen([hostPath UTF8String], RTLD_LAZY);
+	if (!libmono)
+	{
+		NSLog(@"Failed to load libmonosgen-2.0.dylib: %s\n", dlerror());
+		return 1;
+	}
+
+	mono_main _mono_main = (mono_main)dlsym(libmono, "mono_main");
+	if (!_mono_main)
+	{
+		NSLog(@"Could not load mono_main(): %s\n", dlerror());
+		return 1;
+	}
+
+	// Insert hostpath, dll and modId as arguments. Overwrite the first argument which was used to launch this application.
+	for (int i = argc; i > 0; i--)
+		argv[i + 1] = argv[i - 1];
+
+	argv[0] = (char*)[hostPath UTF8String];
```

Comment:
These in general wouldn't be safe (`UTF8String` returns a pointer to an internal buffer that will be overwritten and/or released, which is why the types don't match without the cast), but since we're immediately calling into `_mono_main` I *think* it will be safe.

Reply:
It might be better to just rewrite it in regular C so these conversions wouldn't be necessary. But I won't be doing that.